### DISCUSSION TITLE
Is there any way that distinguish OpenMP kernels from host code?

### DISCUSSION THREAD

**Discussion by fxzjshm** [AdaptiveCpp/AdaptiveCpp#741] [Q&A] [answered]
TL;DR: can I use different code paths in the same function for OpenMP kernels and host code?

Detailed version:

For some reason I have to write a wrapper of device pointers (I mean `sycl::malloc_deivce`) that allows reading/writing device memory like regular variables on host. 
The example code is here ( simplified code of [this](https://github.com/fxzjshm/SyclParallelSTL/blob/9e2d3a7bb3b903b0a74fb6c2ca0b3a0a7d5b38c6/include/sycl/helpers/sycl_device_value.hpp) ):
```c++
#include <sycl/sycl.hpp>

template <typename T>
inline T read_device_pointer(const T *ptr, sycl::queue &q) {
  T tmp;
  q.copy(ptr, &tmp, 1).wait();
  return tmp;
}

template <typename T>
inline void write_device_pointer(T *ptr, const T val, sycl::queue &q) {
  q.copy(&val, ptr, 1).wait();
}

sycl::queue &default_queue() {
  thread_local sycl::queue q;
  return q;
}

template <typename T> class device_value {
public:
  device_value(T *ptr_) : ptr(ptr_) {}

#if 0 // defined(__SYCL_DEVICE_ONLY__) || defined(__SYCL_SINGLE_SOURCE__)
  operator const T() const noexcept { return *ptr; }
  device_value<T> &operator=(const T &value) noexcept {
    *ptr = value;
    return *this;
  }
#else
  operator const T() const noexcept {
    return read_device_pointer(ptr, default_queue());
  }
  device_value<T> &operator=(const T &value) noexcept {
    write_device_pointer(ptr, value, default_queue());
    return *this;
  }
#endif // __SYCL_DEVICE_ONLY__ or __SYCL_SINGLE_SOURCE__

  // default copy function isn't what we need.
  device_value<T> &operator=(const device_value<T> &value) noexcept {
    return operator=(T(value));
  }

private:
  T *ptr;
};

int main() {
  const size_t N = 1 << 12;
  sycl::queue q = default_queue();
  int *u = sycl::malloc_device<int>(N, q);

  int T = 10;
  while (T--) {
    q.parallel_for(sycl::range{N}, [=](sycl::item<1> id) {
       device_value<int> val{u + id.get_id()};
       val = 42;
     }).wait();

    std::cout << "Running..." << std::endl;
  }

  sycl::free(u, q);

  return 0;
}
```
(macro in comment is the way I am using now to distinguish device and host, and `main` function is a test)

However, the "device_pointer" is acutally directly accessible when running kernels on CPU, and accessing using `queue` is quite slow (set 0 and 1 to the `#if` directive to see the difference of the interval of log output)

So is there any way to access the "device" memory directly when running on CPU OpenMP kernels, while remains using `queue` in host code (for execution on GPU) ?

**Comment by illuhad** [comment, marked as answer]
Unfortunately, I do not think this is possible because OpenMP kernels *are* host code and compiled as such. hipSYCL actually makes additional gurantees in this case as an extension, e.g. if you are exclusively in the `omp.library-only` compilation flow, you can invoke regular host code including shared libraries from inside your kernels.

**Reply by fxzjshm** [reply]
Thanks for your detailed explanation! Maybe manually setting a macro to control this is needed then...
